internal/venice-client-common/src/main/java/com/linkedin/venice/compression/GzipCompressor.java

```diff
@@ -19,15 +19,30 @@ public GzipCompressor() {
   @Override
   public byte[] compress(byte[] data) throws IOException {
     ByteArrayOutputStream bos = new ByteArrayOutputStream();
-    GZIPOutputStream gos = new GZIPOutputStream(bos);
-    gos.write(data);
-    gos.close();
-    return bos.toByteArray();
+    try (GZIPOutputStream gos = new GZIPOutputStream(bos)) {
+      gos.write(data);
+      gos.finish();
+      return bos.toByteArray();
+    }
   }
 
   @Override
-  public ByteBuffer compress(ByteBuffer data) throws IOException {
-    return ByteBuffer.wrap(compress(ByteUtils.extractByteArray(data)));
+  public ByteBuffer compress(ByteBuffer data, int startPositionOfOutput) throws IOException {
+    ByteArrayOutputStream bos = new ByteArrayOutputStream();
+    for (int i = 0; i < startPositionOfOutput; i++) {
+      bos.write(0);
+    }
+    try (GZIPOutputStream gos = new GZIPOutputStream(bos)) {
+      if (data.hasArray()) {
+        gos.write(data.array(), data.position(), data.remaining());
+      } else {
+        gos.write(ByteUtils.extractByteArray(data));
```

Comment:
Sure, I can try that tomorrow, although note that this path is for direct BBs, which we don't use ATM, so even if it's more optimized it will not have an impact.

Reply:
I see.
I saw `ByteBuffer` is an abstract class, which may have other implementations, and some of them can be backed by multiple arrays. So it is possible that `hasArray` return false, but it is still a heap-based buffer.

Reply:
I'm looking at the code and I think the proposed change doesn't look super interesting... `GZIPOutputStream` does not implement its own `write(int b)` function but rather uses the one from its parent class, `DeflaterOutputStream`, which is implemented like this:

```
    public void write(int b) throws IOException {
        byte[] buf = new byte[1];
        buf[0] = (byte)(b & 0xff);
        write(buf, 0, 1);
    }
```

So if I do something like:

```
        while (data.hasRemaining()) {
          gos.write(data.get());
        }
```

Then I will end up instantiating one byte[] of length 1 per byte to copy out of the DirectBB... I think it's better to instantiate a single byte[] of the correct length rather than one byte[] per byte... I'll mark this thread as resolved, but feel free to re-open it if you have other comments on this.

Reply:
Sounds good to me.